### PR TITLE
Add seek/pos to buffer unpacker

### PR DESCRIPTION
**Pull request by sskaje**
In my case, I got a buff, which has several msgpack-ed parts.

So I added seek/pos to my fork, and I write a simple script like:

```php
<?php
    $unpacker = new BufferUnpacker($content);
    $unpacker->seek(4);

    while ($unpacker->pos() >= 0) {
       $value = $unpacker->unpack();
        echo "=== Position=" . $unpacker->pos() . " ===\n";
        var_dump($value);
  }

```

I was considering add a fseek() like seek(), but feeling not very useful to me, so I just added a simple one.

### CODE REVIEW

**Comment by rybakit** [comment]
> In my case, I got a buff, which has several msgpack-ed parts.

I wonder if [tryUnpack()](https://github.com/rybakit/msgpack.php#unpacking) would work for your use case? I implemented seeking functionality time ago (check [this](https://github.com/rybakit/msgpack.php/blob/fd9f114d2f27b815dc8b88000fe9453c51c1132a/src/BufferUnpacker.php#L85-L90) and [this](https://github.com/rybakit/msgpack.php/blob/fd9f114d2f27b815dc8b88000fe9453c51c1132a/tests/Unit/BufferUnpackerTest.php#L161-L166)), but removed later as I didn't find a use case for it.

**Comment by sskaje** [comment]
@rybakit IMO tryUnpack() works for:
  1 streaming and then unpack
  2 test for unknown structure

My case looks like

| Packed ID, int32 BE | Command Byte | Flags Byte | Length, int16 BE | 
| msgpack data ...|

It's a known structure, tryUnpack() is not a good choice for me. 
Without fseek() like function, I need to $unpacker->unpack(substr($raw_packet, $offset))
If there are many segments of msgpacked data, with known structure, tryUnpack() is not a good choice.

**Comment by rybakit** [comment]
@sskaje OK, adding `seek()` makes sense then, but I'm not sure about `pos()`. Looking at your example, I do not see how that loop can be terminated, `pos()` will never return `-1` until you put `seek()` inside the loop. From the other hand, I guess 

```php
$parts = $unpacker->seek(4)->tryUnpack();
```
will do the trick. Am I correct?

**Comment by sskaje** [comment]
@rybakit You're right, I'll let seek returns $this and remove pos()

File: src/BufferUnpacker.php [hunk 1]
```diff
@@ -70,6 +70,12 @@ public function reset(string $buffer = '') : self
         return $this;
     }
 
+    public function seek($offset)
```
Comment (rybakit) [inline src/BufferUnpacker.php:73]: Could you change the signature to `public function seek(int $offset) : self`?

File: src/BufferUnpacker.php [hunk 2]
```diff
@@ -70,6 +70,12 @@ public function reset(string $buffer = '') : self
...
+    {
+        $this->offset = (int) $offset;
```
Comment (rybakit) [inline src/BufferUnpacker.php:75]: `(int)` casting can be removed (see above comment).

File: src/BufferUnpacker.php [hunk 3]
```diff
@@ -70,6 +70,12 @@ public function reset(string $buffer = '') : self
...
+        return $this;
```
Comment (rybakit) [inline src/BufferUnpacker.php:76]: Missing blank line, see https://travis-ci.org/rybakit/msgpack.php/jobs/392785445#L1577.

**Comment by rybakit** [comment]
@sskaje I've fixed CS issues, added tests and seek origin argument to `seek()`: https://github.com/rybakit/msgpack.php/commit/0d2818c7121e669952cbaa5437bbeebff2af9743#diff-2a07a1ef20556a8548b7b52903182b2f.

**Comment by rybakit** [comment]
And the [final version](https://github.com/rybakit/msgpack.php/releases/tag/v0.5.3): https://github.com/rybakit/msgpack.php/blob/2c2f046ad6f2cc013770487517eaa99a6fec102a/src/BufferUnpacker.php#L73-L99

**Comment by sskaje** [comment]
I just came back from vacation. 
You did a very good job, thank you.

**Pull request merged**